qiskit/pulse/builder.py

```diff
@@ -0,0 +1,1963 @@
+
+    # Execute on a real backend.
+    # job = execute(pulse_prog, backend)
+
+The builder initializes a :class:`pulse.Schedule`, ``pulse_prog``
+and then begins to construct the program within the context. The output pulse
+schedule will survive after the context is exited and can be executed like a
+normal Qiskit program ``qiskit.execute(pulse_prog, backend)``.
+
+Pulse programming has a simple imperative style. There is little need to deal
+with classes such as the :class:`~qiskit.circuit.QuantumCircuit` or
+:class:`pulse.Schedule`\s. The pulse builder dynamically constructs these
+intermediate representations (IR) behind the scenes. This leaves the programmer
+to worry about the raw experimental physics of pulse programming and not
+constructing cumbersome data structures.
+
+We can optionally pass a :class:`~qiskit.providers.BaseBackend` to
+:func:`build` to enable enhanced functionality. Below, we prepare a Bell state
+by automatically compiling the required pulses from their gate-level
+representations, while simultaneously applying a long decoupling pulse to a
+neighboring qubit. We terminate the experiment with a measurement to see which
+state we prepared. This program which mixes circuits and pulses will be
+automatically lowered to be run as a pulse program:
+
+.. jupyter-execute::
+
+    import math
+
+    from qiskit import pulse
+    from qiskit.test.mock import FakeOpenPulse3Q
+
+    # TODO: This example should use a real mock backend.
+    backend = FakeOpenPulse3Q()
+
+    d2 = pulse.DriveChannel(2)
+
+    with pulse.build(backend) as bell_prep:
+        pulse.u2(0, math.pi, 0)
+        pulse.cx(0, 1)
+
+    with pulse.build(backend) as decoupled_bell_prep_and_measure:
+        # We call our bell state preparation schedule constructed above.
+        with pulse.align_right():
+            pulse.call(bell_prep)
+            pulse.play(pulse.Constant(bell_prep.duration, 0.02), d2)
+            pulse.barrier(0, 1, 2)
+            registers = pulse.measure_all()
+
+    decoupled_bell_prep_and_measure.draw()
+
+With the pulse builder we are able to blend programming on qubits and channels.
+While, the pulse IR of Qiskit is based on instructions that operate on
+channels, the pulse builder automatically handles the mapping from qubits to
+channels for you.
+
+In the example below we demonstrate some more features of the pulse builder:
+
+.. jupyter-execute::
+
+    import math
+
+    from qiskit import QuantumCircuit
+    import qiskit.pulse as pulse
+    import qiskit.pulse.pulse_lib as pulse_lib
+    from qiskit.test.mock import FakeOpenPulse2Q
+
+    backend = FakeOpenPulse2Q()
+
+    with pulse.build(backend) as pulse_prog:
+        # Create a pulse.
+        gaussian_pulse = pulse_lib.gaussian(10, 1.0, 2)
+        # Get the qubit's corresponding drive channel from the backend.
+        d0 = pulse.drive_channel(0)
+        d1 = pulse.drive_channel(1)
+        # Play a pulse at t=0.
+        pulse.play(gaussian_pulse, d0)
```

Comment:
I'm not sure I like incorporating strings. The second one while I like the syntax, off the top of my head it would require pre-generating these channels which could hurt import time and lead to weird global behavior. I think we should punt this down the road and consider as a separate feature.

Reply:
What about removing the `_channel` methods? I see you tend towards `d0 = pulse.DriveChannel(0)`